### PR TITLE
Optimize header includes (8/N)

### PR DESCRIPTION
**Pull request by mvieth**
Continuing #4577 and others.

### CODE REVIEW

File: apps/src/openni_klt.cpp [hunk 1]
```diff
@@ -41,9 +41,9 @@
 #include <pcl/io/pcd_io.h>
 #include <pcl/keypoints/harris_2d.h>
 #include <pcl/tracking/pyramidal_klt.h>
-#include <pcl/visualization/boost.h>
 #include <pcl/visualization/image_viewer.h>
 
+#include <boost/date_time/posix_time/posix_time.hpp>
```
Comment (kunaltyagi) [inline apps/src/openni_klt.cpp:46]: Does this work for Windows system as well?

File: common/include/pcl/common/common_headers.h [hunk 2]
```diff
@@ -35,6 +35,7 @@
 
 #pragma once
 
+PCL_DEPRECATED_HEADER(1, 15, "Please include any needed headers from pcl/common directly.")
```
Comment (kunaltyagi) [inline common/include/pcl/common/common_headers.h:38]: We can let others include the common headers, no issues. OpenCV also provides similar functionality of per-module headers

File: doc/tutorials/content/sources/iros2011/src/build_all_object_models.cpp [hunk 3]
```diff
@@ -7,6 +7,7 @@
 #include <pcl/io/pcd_io.h>
 #include <pcl/visualization/pcl_visualizer.h>
 #include <boost/filesystem.hpp>
+#include <boost/algorithm/string.hpp>
```
Comment (kunaltyagi) [inline doc/tutorials/content/sources/iros2011/src/build_all_object_models.cpp:10]: // for? :)

File: io/include/pcl/io/impl/auto_io.hpp [hunk 4]
```diff
@@ -41,11 +41,11 @@
 #define PCL_IO_AUTO_IO_IMPL_H_
 
 // #include <pcl/io/file_io.h>
-// #include <pcl/io/boost.h>
 #include <pcl/io/pcd_io.h>
 #include <pcl/io/ply_io.h>
 #include <pcl/io/ifs_io.h>
 // #include <pcl/io/vtk_io.h>
+#include <boost/filesystem.hpp> // for path
```
Comment (kunaltyagi) [inline io/include/pcl/io/impl/auto_io.hpp:48]: What about the commented lines?

File: doc/tutorials/content/sources/iros2011/src/build_all_object_models.cpp (hunk 3 above)
Comment (mvieth) [reply doc/tutorials/content/sources/iros2011/src/build_all_object_models.cpp]: Added.

File: io/include/pcl/io/impl/auto_io.hpp (hunk 4 above)
Comment (mvieth) [reply io/include/pcl/io/impl/auto_io.hpp:46]: Removed. Not sure what they were for initially, but they are obviously not needed now.

File: common/include/pcl/common/common_headers.h (hunk 2 above)
Comment (mvieth) [reply common/include/pcl/common/common_headers.h]: I would always recommend to only include the headers that are actually needed, but I removed the deprecation notice again.

File: apps/src/openni_klt.cpp (hunk 1 above)
Comment (mvieth) [reply apps/src/openni_klt.cpp:46]: What do you mean? Why wouldn't it?

**Review by kunaltyagi — COMMENTED** [review]
65 done. Hopefully, you don't mind me reviewing draft ones :)

**Comment by mvieth** [comment]
No problem, I actually consider it ready for review :smile:

File: apps/src/openni_klt.cpp (hunk 1 above)
Comment (kunaltyagi) [reply apps/src/openni_klt.cpp:46]: I didn't know if Posix time is meant to be a front-end header (API) or back-end header (implementation). Just checked that it's a front end header :)

Comment (mvieth) [reply apps/src/openni_klt.cpp:46]: I also find the header structure of the boost libraries confusing. Often, there is no clear information what header should be included to be able to use a specific function (or I just didn't find such information :laughing: ). And it is also not always clear whether a header `<boost/X.hpp>` is a convenience meta-header and it is ok to include a more specific header `<boost/X/Y.hpp>`

**Review by kunaltyagi — COMMENTED** [review]
Some tools have new includes without a `// for ` line, while others have the same line annotated with the reason. Would it be possible to make the changeset a more uniform?

**Comment by mvieth** [comment]
> Some tools have new includes without a `// for ` line, while others have the same line annotated with the reason. Would it be possible to make the changeset a more uniform?

Sure, let me know in case I missed any.

File: search/include/pcl/search/pcl_search.h [hunk 5]
```diff
@@ -38,7 +38,7 @@
  */
 
 #pragma once
-
+PCL_DEPRECATED_HEADER(1, 15, "Please include any needed headers from pcl/search directly.")
```
Comment (kunaltyagi) [inline search/include/pcl/search/pcl_search.h:41]: No need to deprecate this. Just like the one in common

Comment (mvieth) [reply search/include/pcl/search/pcl_search.h]: Alright, removed it.

File: geometry/include/pcl/geometry/mesh_indices.h [hunk 6]
```diff
@@ -40,7 +40,7 @@
 
 #pragma once
 
-#include <pcl/geometry/boost.h>
+#include <boost/operators.hpp>
```
Comment (larshg) [inline geometry/include/pcl/geometry/mesh_indices.h:43]: Is this for totally_orderered, unit_steppable and additive?

**Review by larshg — APPROVED** [review]

File: geometry/include/pcl/geometry/mesh_indices.h (hunk 6 above)
Comment (mvieth) [reply geometry/include/pcl/geometry/mesh_indices.h:43]: Yes, exactly.

**Comment by mvieth** [comment]
BTW this reduces the build time of PCL by about 7% with GCC 9.3.0 on Ubuntu 20.04 (28.2 min before, 26.3 min now) and also seems to reduce the build time of projects that use PCL

**Review by kunaltyagi — APPROVED** [review]
I had reviewed all files (GitHub shows no new changes).

🚀 (10% is huge)

**Comment by mvieth** [comment]
I squashed the 8 commits into 3 commits and force-pushed, which dismissed your approvals. But since I did not make any code changes, I am going to assume that you still approve and merge.

**Pull request merged**